File: experiments/source/ocaml/clipped_huber_sum_reordered.ml
open Loom

let[@loom.entry] clipped_huber_sum_reordered (delta : float) (cap : float)
    (x : Tensor1.t) (y : Tensor1.t) =
  let contribution xi yi =
    let d = xi -. yi in
    let abs_d = if d < 0.0 then 0.0 -. d else d in
    let half_delta = 0.5 *. delta in
    let quadratic = (d *. d) *. 0.5 in
    let linear = delta *. (abs_d -. half_delta) in
    let value = if delta < abs_d then linear else quadratic in
    if cap < value then cap else value
  in
  Tensor1.reduce_sum (Tensor1.map2 contribution x y)

File: experiments/source/python/clipped_huber_sum_reordered.py
import loom


@loom.entry
def clipped_huber_sum_reordered(delta: float, cap: float, x: loom.Tensor1, y: loom.Tensor1) -> float:
    def contribution(xi: float, yi: float) -> float:
        d = xi - yi
        abs_d = 0.0 - d if d < 0.0 else d
        half_delta = 0.5 * delta
        quadratic = (d * d) * 0.5
        linear = delta * (abs_d - half_delta)
        value = linear if delta < abs_d else quadratic
        return cap if cap < value else value

    return loom.Tensor1.reduce_sum(loom.Tensor1.map2(contribution, x, y))
